Worker dependency issue

Looks like there is still a linger issue related to missing dependencies. Here are the repro steps.

Clone https://github.com/aspect-build/bazel-examples

```
cd jest
git checkout open_wc_repro_part_1
bazel clean
bazel build //src:src_test_ts
```
should fails with
```
error TS2307: Cannot find module '@open-wc/dedupe-mixin' or its corresponding type declarations.
```
since there is a missing transitive dependency in an npm package. Now

```
git checkout open_wc_repro_part_2
bazel build //src:src_test_ts
```
should still fail even tho the transitive dependency is now added
```
bazel shutdown
bazel build //src:src_test_ts
```
should now build successfully

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -134,6 +134,10 @@ jobs:
 
             - run: go install github.com/bazelbuild/buildtools/buildozer@latest
               if: ${{ hashFiles(format('{0}/test.sh', matrix.folder)) != '' }}
+            
+            - uses: pnpm/action-setup@v2
+              with:
+                version: '7.14.2'
 
             - name: run ./test.sh
               working-directory: ${{ matrix.folder }}
@@ -142,7 +146,7 @@ jobs:
               run: |
                   cp -f $GITHUB_WORKSPACE/.github/workflows/ci.bazelrc .bazelrc.test
                   echo "build --config=${{ matrix.config }}" >> .bazelrc.test
-                  echo "build --disk_cache=" >> .bazelrc.test
+                  echo "build --disk_cache=/tmp/cache" >> .bazelrc.test
                   bazel clean
                   ./test.sh
                   ./test.sh
@@ -154,7 +158,7 @@ jobs:
               run: |
                   cp -f $GITHUB_WORKSPACE/.github/workflows/ci.bazelrc .bazelrc.test
                   echo "build --config=${{ matrix.config }}" >> .bazelrc.test
-                  echo "build --disk_cache=" >> .bazelrc.test
+                  echo "build --disk_cache=/tmp/cache" >> .bazelrc.test
                   echo "build --noexperimental_allow_unresolved_symlinks" >> ".bazelrc.test"
                   bazel clean
                   ./test.sh

--- a/e2e/worker/pnpm-workspace.yaml
+++ b/e2e/worker/pnpm-workspace.yaml
@@ -1,0 +1,4 @@
+
+packages:
+    - 'transitive_closure/packages/*'
+    - 'transitive_closure'

--- a/e2e/worker/test.sh
+++ b/e2e/worker/test.sh
@@ -154,6 +154,25 @@ message "# Case 12: .tsbuildinfo file should be written when analysis cache is d
 bazel build //composite || exit_with_message "Case 12: expected worker to build without errors."
 bazel build //composite --action_env=ANALYSIS_CACHE_BUST=1 || exit_with_message "Case 12: expected worker to build without errors. (subsequent)"
 
+
+message "# Case 13: should not ignore new transitive deps"
+
+message="error TS2307: Cannot find module '@transitive_closure/b' or its corresponding type declarations."
+bazel build //transitive_closure 2>&1 | grep "$message" || exit_with_message "Case 13: expected worker to report \"$message\""
+
+lockfile_backup=$(mktemp)
+cat pnpm-lock.yaml > $lockfile_backup
+add_trap "cat $lockfile_backup > pnpm-lock.yaml"
+
+
+tmp=$(mktemp -d)
+jq '.pnpm.packageExtensions["@transitive_closure/a"].dependencies["@transitive_closure/b"] = "workspace:*"' package.json > "$tmp/package.json"
+pnpm install --lockfile-only --dir "$tmp" --store-dir $(mktemp -d) --virtual-store-dir $(mktemp -d)
+cat "$tmp/pnpm-lock.yaml" > pnpm-lock.yaml
+
+bazel build //transitive_closure 2>&1  | grep "$message" && exit_with_message "Case 13: expected worker to not report \"$message\""
+
+
 message "# Case X: Should dump traces"
 bazel build //trace
 

--- a/e2e/worker/transitive_closure/BUILD.bazel
+++ b/e2e/worker/transitive_closure/BUILD.bazel
@@ -1,0 +1,14 @@
+load("@aspect_rules_ts//ts:defs.bzl", "ts_project")
+
+load("@npm//:defs.bzl", "npm_link_all_packages")
+npm_link_all_packages(name = "node_modules")
+
+ts_project(
+    name = "transitive_closure",
+    srcs = ["index.ts"],
+    deps = [
+        ":node_modules/@transitive_closure/a",
+    ],
+    tsconfig = "//:tsconfig",
+    tags = ["manual"]
+)

--- a/e2e/worker/transitive_closure/index.ts
+++ b/e2e/worker/transitive_closure/index.ts
@@ -1,0 +1,3 @@
+import { a } from '@transitive_closure/a';
+
+console.log(a);

--- a/e2e/worker/transitive_closure/package.json
+++ b/e2e/worker/transitive_closure/package.json
@@ -1,0 +1,6 @@
+{
+    "name": "transitive_closure",
+    "dependencies": {
+        "@transitive_closure/a": "workspace:*"
+    }
+}

--- a/e2e/worker/transitive_closure/packages/a/BUILD
+++ b/e2e/worker/transitive_closure/packages/a/BUILD
@@ -1,0 +1,21 @@
+load("@aspect_rules_js//js:defs.bzl", "js_library")
+load("@aspect_rules_js//npm:defs.bzl", "npm_package")
+
+load("@npm//:defs.bzl", "npm_link_all_packages")
+npm_link_all_packages()
+
+js_library(
+    name = "a_lib",
+    srcs = [
+        "index.d.ts",
+        "index.js",
+    ]
+)
+
+npm_package(
+    name = "a",
+    srcs = [
+        ":a_lib",
+    ],
+    visibility = ["//visibility:public"],
+)

--- a/e2e/worker/transitive_closure/packages/a/index.d.ts
+++ b/e2e/worker/transitive_closure/packages/a/index.d.ts
@@ -1,0 +1,3 @@
+import {b} from "@transitive_closure/b";
+
+export const a: typeof b;

--- a/e2e/worker/transitive_closure/packages/a/index.js
+++ b/e2e/worker/transitive_closure/packages/a/index.js
@@ -1,0 +1,3 @@
+const b = require("@transitive_closure/b");
+
+module.exports.a = b.b;

--- a/e2e/worker/transitive_closure/packages/a/package.json
+++ b/e2e/worker/transitive_closure/packages/a/package.json
@@ -1,0 +1,4 @@
+{
+    "name": "@transitive_closure/a",
+    "dependencies": {}
+}

--- a/e2e/worker/transitive_closure/packages/b/BUILD
+++ b/e2e/worker/transitive_closure/packages/b/BUILD
@@ -1,0 +1,22 @@
+load("@aspect_rules_js//js:defs.bzl", "js_library")
+load("@aspect_rules_js//npm:defs.bzl", "npm_package")
+
+load("@npm//:defs.bzl", "npm_link_all_packages")
+npm_link_all_packages()
+
+
+js_library(
+    name = "b_lib",
+    srcs = [
+        "index.d.ts",
+        "index.js",
+    ]
+)
+
+npm_package(
+    name = "b",
+    srcs = [
+        ":b_lib",
+    ],
+    visibility = ["//visibility:public"],
+)

--- a/e2e/worker/transitive_closure/packages/b/index.d.ts
+++ b/e2e/worker/transitive_closure/packages/b/index.d.ts
@@ -1,0 +1,1 @@
+export const b: string;

--- a/e2e/worker/transitive_closure/packages/b/index.js
+++ b/e2e/worker/transitive_closure/packages/b/index.js
@@ -1,0 +1,1 @@
+module.exports.b = "1";

--- a/e2e/worker/transitive_closure/packages/b/package.json
+++ b/e2e/worker/transitive_closure/packages/b/package.json
@@ -1,0 +1,3 @@
+{
+    "name": "@transitive_closure/b"
+}

--- a/e2e/worker/tsconfig.json
+++ b/e2e/worker/tsconfig.json
@@ -1,6 +1,10 @@
 {
     "compilerOptions": {
         "module": "ES2020",
-        "moduleResolution": "node"
+        "moduleResolution": "node",
+        "lib": [
+            "ES2015",
+            "DOM"
+        ]
     }
 }

--- a/ts/private/ts_project_worker.js
+++ b/ts/private/ts_project_worker.js
@@ -1,9 +1,15 @@
 const fs = require('fs');
 const path = require('path');
 const v8 = require('v8');
-const util = require("util")
 const ts = require('typescript');
 const worker = require('@bazel/worker');
+
+
+// workaround for the issue introduced in https://github.com/microsoft/TypeScript/pull/42095
+if (Array.isArray(ts.ignoredPaths)) {
+    ts.ignoredPaths = ts.ignoredPaths.filter(ignoredPath => ignoredPath != "/node_modules/.")
+}
+
 
 /** Constants */
 const MNEMONIC = 'TsProject';
@@ -78,7 +84,7 @@ function createFilesystemTree(root, inputs) {
         return node;
     }
 
-    function add(p, hash) {
+    function add(p) {
         const parts = path.parse(p);        
         const dirs = parts.dir.split(path.sep).filter(p => p != "");
 
@@ -94,35 +100,23 @@ function createFilesystemTree(root, inputs) {
             node = node[part];
         }
 
-        // Digest is empty when the input is a symlink which we use as an indicator to limit number of realpath calls made.
-        // See: https://github.com/bazelbuild/bazel/pull/14002#issuecomment-977790796 for why it can be empty.
-        if (hash == null) {
-            const realpath = ts.sys.realpath(path.join(root, p));
-            const relative = path.relative(root, realpath);
-            if (relative != p) {
-                node[parts.base] = {
-                    [Type]: TYPE.SYMLINK,
-                    [Symlink]: relative
-                }
+        const absolutePath = path.join(root, p)
+        const stat = fs.lstatSync(absolutePath)
+
+        if (stat.isSymbolicLink()) {
+            const linkpath = fs.readlinkSync(absolutePath);
+            const absLinkpath = path.isAbsolute(linkpath) ? linkpath : path.resolve(path.dirname(absolutePath, linkpath))
+            const relative = path.relative(root, absLinkpath)
+            node[parts.base] = {
+                [Type]: TYPE.SYMLINK,
+                [Symlink]: relative
             }
             notifyWatchers(dirs, parts.base, TYPE.SYMLINK, EVENT_TYPE.ADDED);
         } else if (parts.base) {
-            const new_node = {
+            node[parts.base] = {
                 [Type]: TYPE.FILE,
-            }
-            try {
-                const linkPath = path.join(root, p)
-                const readlink = fs.readlinkSync(linkPath);
-                const targetPath = path.isAbsolute(readlink) ? readlink : path.join(path.dirname(linkPath), readlink)
-                const relative = path.relative(root, targetPath);
-                if (relative != p) {
-                    new_node[Type] = TYPE.SYMLINK
-                    new_node[Symlink] = relative
-                }
-            } catch (e) { /* Can't determine if it's a symlink. move on */ }
-
-            node[parts.base] = new_node;
-            notifyWatchers(dirs, parts.base, new_node[Type], EVENT_TYPE.ADDED);
+            };
+            notifyWatchers(dirs, parts.base, TYPE.FILE, EVENT_TYPE.ADDED);
         }
     }
 

--- a/ts/test/ts_project_worker/fstree.test.js
+++ b/ts/test/ts_project_worker/fstree.test.js
@@ -13,8 +13,16 @@ const worker = require("./ts_project_worker");
 const root = process.env.GTEST_TMP_DIR;
 const tree = worker.createFilesystemTree(root, {});
 
-tree.add("tree/subtree/input.js", "1");
-tree.add("tree/input.js", "2");
+function createFile(p) {
+    p = path.join(root, p);
+    fs.mkdirSync(path.dirname(p), {recursive: true});
+    fs.writeFileSync(p, "");
+}
+
+createFile("tree/subtree/input.js");
+tree.add("tree/subtree/input.js");
+createFile("tree/input.js");
+tree.add("tree/input.js");
 assert.deepStrictEqual(tree.getDirectories("tree"), ["subtree"]);
 assert.deepStrictEqual(tree.readDirectory("tree"), ["subtree", "input.js"]);
 assert.ok(tree.directoryExists("tree"));
@@ -28,14 +36,16 @@ assert.ok(!tree.fileExists("tree"));
 // Symlinks
 fs.mkdirSync(path.join(root, 'not_a_symlink_but_null', 'deep'), {recursive: true});
 fs.writeFileSync(path.join(root, 'not_a_symlink_but_null', 'deep', "input.js"), "");
-tree.add("not_a_symlink_but_null/deep/input.js", null);
+tree.add("not_a_symlink_but_null/deep/input.js");
 assert.deepStrictEqual(tree.readDirectory("not_a_symlink_but_null"), ["deep"])
 
 
-fs.mkdirSync(path.join(root, "symlink"));
+
+createFile("symlink/to/me/input.js")
 fs.symlinkSync(path.join(root, "symlink"), path.join(root, "symlinked"), "dir");
-tree.add("symlink/to/me/input.js", "3");
-tree.add("symlinked", null);
+
+tree.add("symlink/to/me/input.js");
+tree.add("symlinked");
 assert.ok(tree.directoryExists("symlinked/to/me"));
 assert.ok(!tree.fileExists("symlinked/to/me"));
 assert.ok(!tree.fileExists("symlinked"));
@@ -50,7 +60,8 @@ assert.ok(!tree.directoryExists("symlinked"));
 assert.ok(!tree.directoryExists("symlinked/to"));
 
 
-tree.add("symlinked/to/input.js", "1")
+createFile("symlinked/to/input.js")
+tree.add("symlinked/to/input.js")
 assert.ok(tree.directoryExists("symlinked"));
 assert.ok(tree.directoryExists("symlinked/to"));
 assert.ok(tree.fileExists("symlinked/to/input.js"));
@@ -65,16 +76,18 @@ fs.mkdirSync(path.join(root, "sym"));
 fs.symlinkSync(path.join(root, "dir"), path.join(root, "sym", "s1"), "dir");
 fs.symlinkSync(path.join(root, "dir"), path.join(root, "sym", "s2"), "dir");
 
-tree.add("dir/input.js", "1");
-tree.add("sym/to/input.js", "1");
-tree.add("sym/s1", null);
-tree.add("sym/s2", null);
+createFile("dir/input.js");
+tree.add("dir/input.js");
+createFile("sym/to/input.js");
+tree.add("sym/to/input.js");
+tree.add("sym/s1");
+tree.add("sym/s2");
 assert.deepStrictEqual(tree.getDirectories("sym"), ["to", "s1", "s2"])
 
 // Dangling
 fs.mkdirSync(path.join(root, "dangle"));
 fs.symlinkSync(path.join(root, "dangle"), path.join(root, "may_dangle"), "dir");
-tree.add("may_dangle", null)
+tree.add("may_dangle")
 assert.equal(tree.directoryExists("may_dangle"), false)
 assert.equal(tree.directoryExists("may_dangle/underneath"), false)
 assert.equal(tree.fileExists("may_dangle"), false)
@@ -85,8 +98,10 @@ assert.deepEqual(tree.readDirectory("may_dangle/underneath"), [])
 assert.deepEqual(tree.getDirectories("may_dangle/underneath"), [])
 
 // Remove
-tree.add("correct_remove/to/input.js", "1")
-tree.add("correct_remove/to/path/input.js", "1")
+createFile("correct_remove/to/input.js");
+tree.add("correct_remove/to/input.js");
+createFile("correct_remove/to/path/input.js");
+tree.add("correct_remove/to/path/input.js")
 tree.remove("correct_remove/to/path/input.js");
 assert.ok(tree.fileExists("correct_remove/to/input.js"));
 assert.ok(tree.directoryExists("correct_remove/to"));
@@ -96,14 +111,17 @@ assert.ok(!tree.fileExists("correct_remove/to/path/input.js"));
 // Watcher
 let calls = [];
 let clean = tree.watchDirectory("inputs", (p) => calls.push(p));
-tree.add("inputs/1.js", "1");
+createFile("inputs/1.js");
+tree.add("inputs/1.js");
 assert.deepEqual(calls, ["inputs", "inputs/1.js"]);
 clean();
 
 calls = [];
-tree.add("inputs/to/2.js", "1");
+createFile("inputs/to/2.js");
+tree.add("inputs/to/2.js");
 clean = tree.watchDirectory("inputs/to", (p) => calls.push(p));
-tree.add("inputs/to/1.js", "1");
+createFile("inputs/to/1.js");
+tree.add("inputs/to/1.js");
 assert.deepEqual(calls, ["inputs/to/1.js"]);
 clean();
 tree.remove("inputs");
@@ -111,13 +129,15 @@ tree.remove("inputs");
 
 calls = [];
 clean = tree.watchFile("inputs/1.js", (p) => calls.push(p));
-tree.add("inputs/1.js", "1");
+createFile("inputs/1.js");
+tree.add("inputs/1.js");
 assert.deepEqual(calls, ["inputs/1.js"]);
 clean();
 
 
 calls = [];
-tree.add("inputs/1.js", "1");
+createFile("inputs/1.js");
+tree.add("inputs/1.js");
 let clean1 = tree.watchDirectory("inputs", (p) => calls.push(p));
 let clean2 = tree.watchFile("inputs/1.js", (p) => calls.push(p));
 tree.remove("inputs/1.js");
@@ -127,7 +147,8 @@ clean2();
 
 
 calls = [];
-tree.add("inputs/1.js", "1");
+createFile("inputs/1.js");
+tree.add("inputs/1.js");
 clean = tree.watchFile("inputs/1.js", (p) => calls.push(p));
 tree.update("inputs/1.js");
 assert.deepEqual(calls, ["inputs/1.js"]);
@@ -141,7 +162,8 @@ let watcher3 = tree.watchFile("inputs/1.js", (p) => calls.push("watcher3"));
 let watcher4 = tree.watchFile("inputs/1.js", (p) => calls.push("watcher4"));
 watcher1();
 watcher2();
-tree.add("inputs/1.js", "2");
+createFile("inputs/1.js");
+tree.add("inputs/1.js");
 assert.deepEqual(calls, ["watcher3", "watcher4"]);
 watcher3();
 watcher4();
@@ -149,9 +171,12 @@ watcher4();
 
 calls = [];
 clean = tree.watchDirectory("recursive", (p) => calls.push(p), true);
-tree.add("recursive/1.js", "2");
-tree.add("recursive/to/1.js", "2");
-tree.add("recursive/to/deep/1.js", "2");
+createFile("recursive/1.js");
+tree.add("recursive/1.js");
+createFile("recursive/to/1.js");
+tree.add("recursive/to/1.js");
+createFile("recursive/to/deep/1.js");
+tree.add("recursive/to/deep/1.js");
 assert.deepEqual(calls, [
     "recursive", 
     "recursive/1.js", 
@@ -167,7 +192,8 @@ clean();
 calls = [];
 let clean3 = tree.watchDirectory("recursive", (p) => calls.push(p), true);
 let clean4 = tree.watchDirectory("recursive", (p) => calls.push(p));
-tree.add("recursive/1.js", "2");
+createFile("recursive/1.js");
+tree.add("recursive/1.js");
 assert.deepEqual(calls, [
     "recursive", 
     "recursive", 

